File: TelegramApp/src/game/gameplay/MissionService.ts
// Ported from Assets/PixelPlanets/StellarBreaker/Scripts/Gameplay/MissionService.cs
import { BigNumber, toBigNumberData, fromBigNumberData, type BigNumberData } from '../core/BigNumber'
import { Emitter } from '../core/Emitter'
import { MissionType, type MissionDefinition } from '../config/MissionDefinition'
import type { CurrencyService } from '../economy/CurrencyService'

/**
 * Fixed one-shot quest list (no rotation/re-roll in this pass). Tracks a running BigNumber
 * counter per mission (planets destroyed / cumulative tap damage / ship upgrades bought),
 * and lets the UI claim a Gold reward once the target is reached. Claimed missions stay
 * claimed forever (persisted).
 */
export class MissionService {
  private readonly defs: MissionDefinition[]
  private readonly wallet: CurrencyService
  private readonly progress: BigNumber[]
  private readonly claimed: boolean[]

  /** (missionIndex) fired when a mission's progress changes. */
  readonly onProgressChanged = new Emitter<number>()
  /** (missionIndex, goldPaid) fired when a mission is claimed. */
  readonly onClaimed = new Emitter<{ index: number; gold: BigNumber }>()

  constructor(defs: MissionDefinition[], wallet: CurrencyService) {
    this.defs = defs
    this.wallet = wallet
    this.progress = new Array(defs.length).fill(BigNumber.Zero)
    this.claimed = new Array(defs.length).fill(false)
  }

  get count(): number {
    return this.defs.length
  }
  def(i: number): MissionDefinition {
    return this.defs[i]
  }
  progressOf(i: number): BigNumber {
    return this.progress[i]
  }
  isClaimed(i: number): boolean {
    return this.claimed[i]
  }
  isComplete(i: number): boolean {
    return this.progress[i].gte(new BigNumber(this.defs[i].target))
  }

  /** 0..1, clamped - for progress bars. */
  progress01(i: number): number {
    const target = Math.max(1e-9, this.defs[i].target)
    const frac = this.progress[i].toNumber() / target
    return Math.min(1, Math.max(0, frac))
  }

  claim(i: number): boolean {
    if (i < 0 || i >= this.defs.length) return false
    if (this.claimed[i] || !this.isComplete(i)) return false
    this.claimed[i] = true
    const reward = new BigNumber(this.defs[i].goldReward)
    this.wallet.add(reward)
    this.onClaimed.emit({ index: i, gold: reward })
    return true
  }

  // -- event hooks (wired by GameSession) --
  notifyPlanetDestroyed(): void {
    this.addProgress(MissionType.DestroyPlanets, BigNumber.One)
  }
  notifyTapDamage(dmg: BigNumber): void {
    this.addProgress(MissionType.TapDamageTotal, dmg)
  }
  notifyShipUpgraded(): void {
    this.addProgress(MissionType.ShipUpgrades, BigNumber.One)
  }

  private addProgress(type: MissionType, amount: BigNumber): void {
    for (let i = 0; i < this.defs.length; i++) {
      if (this.defs[i].type !== type || this.claimed[i]) continue
      this.progress[i] = this.progress[i].add(amount)
      this.onProgressChanged.emit(i)
    }
  }

  // -- save/load --
  captureProgress(): BigNumberData[] {
    return this.progress.map(toBigNumberData)
  }

  captureClaimed(): boolean[] {
    return [...this.claimed]
  }

  restoreProgress(progress: BigNumberData[] | undefined | null, claimed: boolean[] | undefined | null): void {
    if (progress) {
      const n = Math.min(progress.length, this.progress.length)
      for (let i = 0; i < n; i++) this.progress[i] = fromBigNumberData(progress[i])
    }
    if (claimed) {
      const n = Math.min(claimed.length, this.claimed.length)
      for (let i = 0; i < n; i++) this.claimed[i] = claimed[i]
    }
  }
}

File: Unity/Assets/PixelPlanets/StellarBreaker/Scripts/Gameplay/MissionService.cs
using System;
using System.Collections.Generic;
using StellarBreaker.Core;
using StellarBreaker.Config;
using StellarBreaker.Economy;

namespace StellarBreaker.Gameplay
{
    /// <summary>
    /// Fixed one-shot quest list (no rotation/re-roll in this pass). Tracks a running BigNumber
    /// counter per mission (planets destroyed / cumulative tap damage / ship upgrades bought),
    /// and lets the UI claim a Gold reward once the target is reached. Claimed missions stay
    /// claimed forever (persisted).
    /// </summary>
    public class MissionService
    {
        readonly IReadOnlyList<MissionDefinition> _defs;
        readonly CurrencyService _wallet;
        readonly BigNumber[] _progress;
        readonly bool[]      _claimed;

        /// <summary>(missionIndex) fired when a mission's progress changes.</summary>
        public event Action<int> OnProgressChanged;
        /// <summary>(missionIndex, goldPaid) fired when a mission is claimed.</summary>
        public event Action<int, BigNumber> OnClaimed;

        public MissionService(IReadOnlyList<MissionDefinition> defs, CurrencyService wallet)
        {
            _defs   = defs   ?? throw new ArgumentNullException(nameof(defs));
            _wallet = wallet ?? throw new ArgumentNullException(nameof(wallet));
            _progress = new BigNumber[_defs.Count];
            _claimed  = new bool[_defs.Count];
            for (int i = 0; i < _progress.Length; i++) _progress[i] = BigNumber.Zero;
        }

        public int Count => _defs.Count;
        public MissionDefinition Def(int i) => _defs[i];
        public BigNumber Progress(int i)    => _progress[i];
        public bool IsClaimed(int i)        => _claimed[i];
        public bool IsComplete(int i)       => _progress[i] >= new BigNumber(_defs[i].target);

        /// <summary>0..1, clamped — for progress bars.</summary>
        public float Progress01(int i)
        {
            double target = Math.Max(1e-9, _defs[i].target);
            double frac = _progress[i].ToDouble() / target;
            return (float)Math.Min(1.0, Math.Max(0.0, frac));
        }

        public bool Claim(int i)
        {
            if (i < 0 || i >= _defs.Count) return false;
            if (_claimed[i] || !IsComplete(i)) return false;
            _claimed[i] = true;
            var reward = new BigNumber(_defs[i].goldReward);
            _wallet.Add(reward);
            OnClaimed?.Invoke(i, reward);
            return true;
        }

        // ── event hooks (wired by GameSession) ──
        public void NotifyPlanetDestroyed()    => AddProgress(MissionType.DestroyPlanets, BigNumber.One);
        public void NotifyTapDamage(BigNumber dmg) => AddProgress(MissionType.TapDamageTotal, dmg);
        public void NotifyShipUpgraded()       => AddProgress(MissionType.ShipUpgrades, BigNumber.One);

        void AddProgress(MissionType type, BigNumber amount)
        {
            for (int i = 0; i < _defs.Count; i++)
            {
                if (_defs[i].type != type || _claimed[i]) continue;
                _progress[i] = _progress[i] + amount;
                OnProgressChanged?.Invoke(i);
            }
        }

        // ── save/load ──
        public BigNumberData[] CaptureProgress()
        {
            var arr = new BigNumberData[_progress.Length];
            for (int i = 0; i < arr.Length; i++) arr[i] = BigNumberData.From(_progress[i]);
            return arr;
        }

        public bool[] CaptureClaimed() => (bool[])_claimed.Clone();

        public void RestoreProgress(BigNumberData[] progress, bool[] claimed)
        {
            if (progress != null)
            {
                int n = Math.Min(progress.Length, _progress.Length);
                for (int i = 0; i < n; i++) _progress[i] = progress[i].To();
            }
            if (claimed != null)
            {
                int n = Math.Min(claimed.Length, _claimed.Length);
                for (int i = 0; i < n; i++) _claimed[i] = claimed[i];
            }
        }
    }
}
